### ISSUE TITLE
Include information about the provisioner and upgrade driver in the Cluster

### ISSUE THREAD

**Issue by sc68cal** [Mirantis/Kostyor#92]
Currently, we use configuration to determine what driver is used for upgrades.

With #89 - we may need to consider associating an upgrade driver with a `Cluster` - so that different `Clusters` that are deployed with different methods can co-exist in Kostyor.

Since we already use driver specific discovery methods, perhaps when a cluster is discovered, the discovery method will be linked to the upgrade driver. For instance, using the Ansible driver would mean that the Ansible upgrade driver will be utilized. Puppet based discovery would use the puppet driver, fuel-ccp would use fuel-ccp, etc. etc. etc.

tl;dr - maybe add a `deployment_driver` attribute to the `Cluster` model? I think maybe something similar to what @Settis did in https://github.com/Settis/Kostyor/commit/40daddc219cae5a522b8630bbdea97cbdb7addd3#diff-ff78bed6d4c4080829b02656feff4055R69

**Comment by ikalnytskyi** [comment]
When I was working on `kostyor-openstack-ansible` driver I considered both `Cluster - Upgrade Driver` and `Upgrade Task - Upgrade Driver` associations. At the beginning it seemed reasonable to me to use the first approach but then I've encountered a need to have alternative driver implementation. In my case it was a different way to run Ansible playbooks, but something similar may be actual for other drivers too.

So in my opinion, it makes sense to use @Settis's proposal and associate driver with `Upgrade Task`, not `Cluster`, and if something goes wrong allow to pass another driver if user wants to.

**Comment by sc68cal** [comment]
The only concern I have, is having to pass in the driver every time we do an upgrade. Is it realistic that a cluster will be provisioned with ansible, and be upgraded with a different driver? I think it's intrinsic to the `Cluster` itself.

**Comment by ikalnytskyi** [comment]
@sc68cal sorry for the delay in my response, I missed that one.

I agree that passing the driver every time may be annoying, but I don't see other solution. `kostyor-openstack-ansible` suggests that we do need separation. Here's the points:

- Drivers may be distributed with several implementations. E.g. `kostyor-openstack-ansible` comes with 2 of them:

  ```python
  'kostyor.upgrades.drivers': [
      'openstack-ansible = kostyor_openstack_ansible.upgrades.ref:Driver',
      'openstack-ansible-alt = kostyor_openstack_ansible.upgrades.alt:Driver',
  ],
  ```

  The implementations are almost identical except the way how we run Ansible playbooks. In this very case driver settings could be a solution, but what if someone implement alternative implementation as a third-party package and reuse significant parts of the original one?

- Discover and Upgrade drivers use different namespaces.

  ```python
  'kostyor.upgrades.drivers': [
      'openstack-ansible = kostyor_openstack_ansible.upgrades.ref:Driver',
      'openstack-ansible-alt = kostyor_openstack_ansible.upgrades.alt:Driver',
  ],
  'kostyor.discovery_drivers': [
      'openstack-ansible = kostyor_openstack_ansible.discover:Driver',
  ],
  ```

  Hence, there's no guarantee that their names (IDs) will always be exactly the same.

So I think there definitely should be a way to explicitly pass which driver we want to use. Maybe if we have some an "upgrade task" (which is failed) and we want to rerun it, then we shouldn't require users to pass driver again, but the initial run, in my opinion, should require that.

What do you think?
